### ISSUE TITLE
Add details about `most_recent_task` in the `GET /tasks` endpoint

### ISSUE THREAD

**Issue by benoit74** [openzim/zimfarm#885]
Currently the `GET /tasks` endpoint (used to list tasks) is not returning the `most_recent_task` of the associated schedule.

This presents the drawback that when we need to display last task in the UI for a bunch of tasks, we have to perform one request per schedule to the `GET /schedules/{schedule_name}` endpoint. This is causing issues like #882.

As discussed in https://github.com/kiwix/k8s/issues/149#issuecomment-1851442182, should we change this now that things have matured / changed a bit?

**Comment by rgaudin** [comment]
I'm in favor

**Comment by elfkuzco** [comment]
Given the `Failed` tab of the Pipelines now has buttons to manually load a specific schedule or load all failed tasks, is this still necessary?

**Comment by rgaudin** [comment]
This refresh status button is there exactly because this information is not present. I think it still is necessary

**Comment by benoit74** [comment]
Yes, the "refresh status" button is a kind of hack to avoid overwhelming the API.

It would be better to not need the "refresh status" button and not need extra API calls.

I'm not sure about the perf impact of making such a move, hence the fact this issue is still not implemented. Probably worth to check how long it takes to load 100 tasks withtout most_recent_task details and 100 tasks with most_recent_task details. And probably worth to add a query parameter to make addition of most_recent_task details optional. Maybe worth to make the move in two steps: first add the query parameter + API logic but do not modify UI. Deploy on prod and make some measurements to have real perf impact. If measurements are ok, modify UI, otherwise iterate.

**Comment by elfkuzco** [comment]
I've been implementing the logic to add most recent task in the `/tasks` endpoint but I've been thinking about it for a while and I was wondering if it really should be there from a "separation of concerns" standpoint as a task having a "most-recent-task" entry seems kind of recursive. Generally, API returns list results in the form
```sh
{
    "meta": // pagination-related details,
     "items": [] // list of resource
}
```
Moreover, there could really be no relationship between an entry and it's "most recent task" since everything is going to be in the "items" entry. I understand it's only the UI that needs this and there's not really anywhere it's going to  be used which makes me think maybe the current approach of having the user request for all tasks via the button is probably better. What do you think?

Currently, the UI builds up a unique list of schedule ids and fetches these leading to multiple "most-recent" tasks being populated assuming multiple failed tasks share the same schedule

**Comment by benoit74** [comment]
Reading your comment a second time, I'm not sure we are on the same page.

What I (finally, I know I was hesitant on that) consider we could implement is something like:

```
{
    "meta": // pagination-related details,
     "items": [
        {
            "id": ...,
             // other task attributes
            "schedule_id": ...,
            "schedule_most_recent_task_id": ...,
            "schedule_most_recent_task_status": ...,
            "schedule_most_recent_task_updated_at": ...,
     ] // list of resource
}
```

We hence do not add the "most recent task of this task", but the "most recent task of the schedule of this task". There is no "infinite recursion". And we can probably add a parameter on the query to make these values optional to save resources when we do not need them (which is generally the case).

Is that what you were considering? Can you be more specific in what causes you trouble here?

**Comment by elfkuzco** [comment]
Okay. I get the idea now

**Comment by elfkuzco** [comment]
It's just, that I was thinking that there is no single query that can achieve this and SQLAlchemy `selectinload`/`joinedload` will emit additional queries to fetch the `Schedule`'s most recent task which is equivalent to the button effect we have. And given we have three tabs that all fetch tasks, modifying it for one (with an optional query param) achieves the same effect (just more on the DB side)

EDIT: Never mind, I've gotten it now.

**Comment by benoit74** [comment]
Main issue with anything done UI-side is not much DB load but the additional number of API calls (100 API calls, that's a lot for an API server, especially since it causes in turn 100 DB queries or more). With the approach proposed here impact should be modest since we paginate anyway, so at most we have 100 more schedules and 100 more tasks to load by ID, maybe 2 additional SQL queries if we don't achieve to make it fit into one single query, but that's hopefully "nothing"

**Comment by elfkuzco** [comment]
True. Makes sense.
